Improve docs for 0.3 (fix #51) + changelog + make it nim v2.0 compatible

pietroppeter opened the pull request:
improve documentation
- [x] improve readme
  - [x] add mention of js interactivity
  - [x] current status
  - [x] example sites
  - [x] features
  - [x] minimal howto use (shows nbook init and nbook build)
  - [x] review contribute
  - [x] remove tips and tricks (all content is in docs)
- [x] improve book
  - [x] change toc structure
  - [x] chapter content: nim default template or markdown
    - [x] add latex here and remove latex chapter
  - [x] chapter toc: explain better the toc dsl
  - [x] chapter configuration: list all configuration options
  - [x] chapter cli: explain better the cli
    - [x] how to customize compile flags
    - [x] how to build a single file (use dump)
    - [x] parallel builds and how to switch them off (and how to set max parallel builds)
  - [x] remove chapter tasks (not interesting for the user)

additional improvements of this PR:
- [x] improve nbook cli help (document dump and build options)
- [x] changed behaviours of clean and update commands
- [x] less logging (no config and toc every time you use nbook)
- [x] remove the rust file
- [x] add as vendored the assets (css, font and js stuff)
- [x] remove toml serialization dependency (make it nim 2.0 ready)
- [x] add nim devel and stable to CI

created changelog complete with history of 0.1 and 0.2 + updates on 0.3:
- [x] add latex support with latex as in nimib (PR #52)
- [x] recent parallel build by beef
- [x] fixes to parallel build by Hugo (#61)
- [x] document changes in #50
  - [x] refactoring and code structure (review)
  - [x] templates
  - [x] book.json
  - [x] blocks context in document.mustache
  - [x] cfg stuff
    - [x] note that changes also fix https://github.com/pietroppeter/nimibook/issues/48 (you should be able to change name of book folder)
  - [x] remove default parameters and allow to pass parameters to build
  - [x] improved init
  - [x] new nb/doc object in Book
    - [x] refactored Book.book_title into Book.title. Overrides previously existing Book.title which was supposed to be used for the Chapter title
    - [x] removed book.path_to_root (it is a document/chapter thing)
  - [x] remove rootfolder in toc and path field in Toc
- [x] after nimib 0.3 is released, change nimib dep

create (or link) issues for other improvements to do in next versions:
- [x] add nim code highlighting when starting from md file #64 
- [x] support part title in toc #45 
- [x] support SUMMARY.md #62 
- [x] nbook binary #63
- [x] improve clean #66
- [x] dumped book.json (now a public api) should not contain cfgDir and it should only contain relative dirs #67
- [x] quit failures for check and build should be only in cli #68

**edit**: this first PR message was multiple times edited (in particular adding todo points from #51)
to keep it up to date with PR content. The discussion below might reflect an earlier version of this ,message.

HugoGranstrom commented:
Seems like you are using the `NETLIFY_SITE_ID` from your nimib project instead of the nimibook one

pietroppeter commented:
not sure why this is still hanging here, my bad, will try to look a this. also #51 has a bunch of tasks I listed back then and not sure if they are covered by this (which is essentially just adding a changelog and making sure we use nimib 0.3 - which is now probably outdated already - we should use latest stable...

HugoGranstrom commented:
Yes it would be nice if we could get nimibook 0.3 released soon, so we can focus fully on the SSG + nimib org in the coming months. I will try and get https://github.com/pietroppeter/nimib/pull/163 done this week, so we can merge #61 as soon as possible as well. Is there anything else code-wise that has to be done, or was stabilizing the parallel build the final thing?

pietroppeter commented:
No, I do not recall anything code wise that needed to be done. It was mostly documenting all the changes the went through the refactoring. For the async thing at the time the idea was to disable, now there is no need to do it.

pietroppeter commented:
Does #61 depends on that nimib code thing?

HugoGranstrom commented:
Ok, that's good 👌

HugoGranstrom commented:
I added the parallel out.js fix to that PR 🙃 could have done it in a separate PR of course... But I just have to go through all the existing nimib documents again and confirm nothing has been broken by it. (When you have to read through all of scinim/getting-started you realize that we have quite a bit of material there 🤣)

pietroppeter commented:
> added the parallel out.js fix to that PR

ah ok right, you mentioned that

HugoGranstrom commented:
If I don't find any bugs in it, I should be able to have it done tomorrow evening. And then I should have #61 done on saturday. Just in time before the speaking hours 😎

pietroppeter commented:
wait the error in tests I see is due to the out.js and is the one that will be fixed?

HugoGranstrom commented:
Yes I think so, it's hard to tell without seeing the log file though 😛 I think we are always compiling a js file even if no nbJs is used, but I fixed that one as well in the nimib PR. Hence the error here even if we don't use nbJs

pietroppeter commented:
> wait the error in tests I see is due to the out.js and is the one that will be fixed?

I see that if I run multiple times `nimble genbook` I occasionally get the same error I see in CI and it seems to be tied to `out.js`. Why are we producing a js since this book has no js functionalities?

HugoGranstrom commented:
Because I never added a check to see if we actually wrote any nbJs code. So it just compiles an empty `.nim` file in this case for every file. As I said, I have fixed this as well in the nimib PR by checking if we have any code to compile.

pietroppeter commented:
anyway now I remember what this PR was about, documentation on how to use nimibook needs to be improved substantially. things changed in last refactoring and they are not documented anywhere (except in the code9

HugoGranstrom commented:
Oh... Yeah that's a bit more 😅 I would need to read that tutorial as well 🙊

pietroppeter commented:
finally this PR is ready to be merged! @HugoGranstrom do you want to give a quick look before I merge and tag 0.3 release?

I did change substantially the documentation that now reflects the current state of nimibook.
Also the changelog is now rather complete.
And I have also made some additional improvements in this PR:
- added support for nim 2.0 by removing dependency on toml_serialization (used the useful `loadTomlSection` from nimib)
- changed behaviors of some commands (clean and update)

see first PR message for all the details (I kept it up to date).

HugoGranstrom commented on book/content.nim:
```diff
@@ -1,23 +1,82 @@
+
+nbText: &"""## Inline equations
+
+**Inline equations** are obtained with delimiters `$..$`. The following inline paragraph:
```
This is a bit picky, but I read this as you should write equations like `$. e{\pi i} .$`. Adding a third dot and adding spaces makes it more obvious that the points just placeholders.
```suggestion
**Inline equations** are obtained with delimiters `$ ... $`. The following inline paragraph:
```

HugoGranstrom commented on book/content.nim:
```diff
@@ -1,23 +1,82 @@
+
+## Block equations
+
+**Block equations** are obtained with delimiters `$$..$$`. The follwing block equation:
```
```suggestion
**Block equations** are obtained with delimiters `$$ ... $$`. The follwing block equation:
```

HugoGranstrom reviewed (changes requested):
Awesome work on cleaning up and writing documentation! :smile: 

I didn't find anything major, just a few link-related comments that needs to be fixed and a very picky comment on the Latex part :)

pietroppeter commented on README.md:
```diff
@@ -3,101 +3,124 @@
+It still has some features missing from mdbook
+(see [this issue](https://github.com/pietroppeter/nimibook/issues/9#issuecomment-851989939)).
+
+To follow up on recent changes check the [changelog.md](changelog.md).
```
Should be fixed now (although it currently links to a file that does not yet exists...)

pietroppeter commented:
Thanks! Will wait for CI to finish and merge after giving another look at the preview.
